app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt

```diff
@@ -1128,6 +1148,58 @@ class HomeFragment : Fragment() {
         binding.sessionControlRecyclerView.adapter?.notifyDataSetChanged()
     }
 
+    private fun onCategoryClick(clickedCategory: PocketRecommendedStoryCategory) {
+        val allCategories = homeFragmentStore.state.pocketStoriesCategories
+
+        // Ensure the number of categories selected at a time is capped.
+        val currentlySelectedCategoriesCount = allCategories.fold(0) { count, category ->
+            if (category.isSelected) count + 1 else count
+        }
+        val oldestCategoryToDeselect = if (
+            currentlySelectedCategoriesCount == POCKET_CATEGORIES_SELECTED_AT_A_TIME_COUNT &&
+            clickedCategory.isSelected.not()
+        ) {
+            allCategories
+                .filter { it.isSelected }
+                .reduce { oldestSelected, category ->
+                    when (oldestSelected.lastInteractedWithTimestamp <= category.lastInteractedWithTimestamp) {
+                        true -> oldestSelected
+                        false -> category
+                    }
+                }
+        } else {
+            null
+        }
+
+        val updatedCategories = allCategories.map {
+            when (it.name) {
+                oldestCategoryToDeselect?.name -> {
+                    it.copy(
+                        isSelected = false,
+                        lastInteractedWithTimestamp = System.currentTimeMillis()
+                    )
+                }
+                clickedCategory.name -> {
+                    it.copy(
+                        isSelected = !it.isSelected,
+                        lastInteractedWithTimestamp = System.currentTimeMillis()
+                    )
+                }
+                else -> it
+            }
+        }
+
+        homeFragmentStore.dispatch(HomeFragmentAction.PocketStoriesCategoriesChange(updatedCategories))
```

Comment:
The actual creation of the new state (the new categories) should preferably happen in the reducer. That way you are guaranteed to always end up with a consistent state. Imagine the processing being slow and the user clicking categories fast: You may generate states here while the new state is not available in the store yet, which will cause inconsistent states. This doesn't not happen in the reducer where the actions will be processed sequentially.

Reply:
Remembered that I tried that first but the reducer seemed a bit big and it seemed to now contain the logic on for how to handle categories being clicked so I then moved that code in composable's parent - this fragment.
https://github.com/mozilla-mobile/fenix/pull/21104#discussion_r700212421

Reply:
I think reducers and state updates happen only on the main thread?
If so, instead of me launching coroutines to calculate the updated state and dispatch that I could have all this running on main.
That would avoid reducers becoming too complex while the new state can be calculated a bit closer to the user interaction.

Reply:
I think actions like "SelectCategory" or "DeselectCategory" make totally sense to me - and I don't think it's a problem if the reducer is complex, if the calculation of the new state is complex (as long it's still a function that just transforms input into an output without side effects). If I remember correctly the first approach had something like a "Clicked" action and that felt wrong because that is too close to the UI layer, the fragment handles the click and would dispatch an actual state changing action (not an UI event like "click").